clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java

```diff
@@ -414,6 +414,61 @@ public boolean matches(AbstractRequest body) {
         assertEquals(singleton(t1p), rebalanceListener.assigned);
     }
 
+    @Test
+    public void testInvalidCoordinatorAssignment() {
```

Comment:
Wrote the test first as it failed with this error just as described in the issue:
```
org.apache.kafka.clients.consumer.internals.ConsumerCoordinatorTest > testInvalidCoordinatorAssignment FAILED
    java.lang.IllegalArgumentException: Assigned partition test2-0 for non-subscribed topic; subscription is [test1]
        at org.apache.kafka.clients.consumer.internals.SubscriptionState.assignFromSubscribed(SubscriptionState.java:195)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.onJoinComplete(ConsumerCoordinator.java:249)
        at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.joinGroupIfNeeded(AbstractCoordinator.java:410)
        at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureActiveGroup(AbstractCoordinator.java:344)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.poll(ConsumerCoordinator.java:340)
        at org.apache.kafka.clients.consumer.internals.ConsumerCoordinatorTest.testInvalidCoordinatorAssignment(ConsumerCoordinatorTest.java:462)

1 test completed, 1 failed
```

Reply:
I think there are two possible scenarios for this case:

1. The coordinator has assigned partitions that we did not ask for.
2. The subscription has changed.

When I filed the bug, it was mostly the second scenario that I was thinking about. If the subscription has changed, then it is reasonable to request another rebalance. However, in the first case, raising the exception may have been a better approach since it likely means a bug in the assignment logic. 

In fact, we keep track of the subscription that was sent in the JoinGroup through `ConsumerCoordinator.joinedSubscription`. Do you think it would make sense to validate the assigned partitions against that set in the case that `assignFromSubscribed` returns false? If there are assigned partitions that we didn't ask for, we can still raise an exception.

Reply:
Done.